source/dub/commandline.d

```diff
@@ -678,7 +678,13 @@ abstract class PackageBuildCommand : Command {
 			}
 		}
 
-		dub.project.validate();
+		auto validity = dub.project.validate();
+
+		if (!validity.matchingSelections) {
+			logInfo("-> Run \"dub upgrade\" to select specified package versions.");
+		}
+
+		enforce(!validity.hasFatalIssues, "Aborting due to fatal issues in project configuration.");
```

Comment:
```suggestion
		enforce(!validity.hasFatalIssues, "Dub.selection.json validation failed.  project configuration.");
```

Though we probably can make the message even better.

Reply:
the validity check isn't just dub.selections.json. For now it is the only reason how a build should be able to fail after validation, but it's designed so that the project validation can be any issue. I can add a reference to that file in the message above though.